### ISSUE TITLE
Extended stack traces

### ISSUE THREAD

**Issue by rahulmutt** [typelead/eta#494]
Due to the very simple nature of Java's `StackTraceElement`, we are unable to store extra information by default which severely limits the debugging information we can display to the user upon throwing an exception. 

I propose that we store the extra debugging information as an annotation and extract it at runtime using reflection when printing the stack trace to give very precise information (including column number ranges).

The `StackTraceElement` gives us the class name, method name, and line number. With the class name we can retrieve the `Class` instance, with the method name, the `Method` instance, and then extract the `FunctionLineInfo` annotation (defined below). The annotation will look as follows:

```java
package eta.runtime.annotation;

public @interface FunctionLineInfo {
  LineInfo[] callStack();
}
```
This annotation stores all the extended line info for a given method (will typically be attached to the `enter` method).

```java
package eta.runtime.annotation;

public @interface LineInfo {
  int lineNumber();
  SourceNote[] sourceNotes();
}
```
Because we have to condense a set of source notes to a single one, we'll end up with a single line number to choose. That single line number that we choose will be the index to the rest of the source notes in the set shown above.

```java
package eta.runtime.annotation;

public @interface SourceNote {
  String functionName();
  SrcSpan srcSpan();
}
```

```java
package eta.runtime.annotation;

public @interface SrcSpan {
  String file();
  int startColumn();
  int startLine();
  int startColumn();
  int startLine();
}
```

So now that we have access to *all* the source notes as an annotation, when an exception is thrown, we can read the `StackTraceElement[]` and render a custom stack trace using all the precise information from above that can yield lots of useful information for the user.

**Comment by rahulmutt** [comment]
@jneira This would be the next step if we wanted to take the debugging info further.

**Comment by jneira** [comment]
It will be amazing that eta has better error reporting than java and [ghc](https://simonmar.github.io/posts/2016-02-12-Stack-traces-in-GHCi.html)

**Comment by rahulmutt** [comment]
Yes that's the goal :) As for the steps required to implement this:

1) Support generation of [RuntimeVisibleAnnotations](https://docs.oracle.com/javase/specs/jvms/se7/html/jvms-4.html#jvms-4.7.16) classfile attribute for `codec-jvm`.
2) Modify `cgTick` to generate the annotations mentioned above. You'll also need some way of telling the codegen monad to include the method-level attributes you're generating.
3) Override `printStackTrace` for `eta.runtime.exception.EtaException` to use reflection to read the attributes above from the stack trace and print out a custom stack trace that just shows modules/functions/line numbers/column numbers in the actual source file format, instead of the class-level encoding that show up in the current method (all the module names lowercased, etc).

**Comment by jneira** [comment]
i have a question: what would be the advantages of store the debug info with annotations over other options (sorry if they are not possible at all):
* Store directly as a static field or method in the class and access them (with reflection or directly?): it could return a map between lines numbers and the rest of info with the structure of your analysis. Or it could be a instance method implementing a interface to make access through virtual calls and no reflection.
* Store as jvm bytecode attributes and use some type of bytecode instrumentation at runtime to read them: is is possible to store arbitrary atributtes in class or methods, the jvm will ignore them.

**Comment by rahulmutt** [comment]
No issues. I sometimes fail to think of the simpler solutions - always good to propose alternatives :) 

1) This solution works too, but the amount of space overhead in generating the data structure storing all this information is more because you have to specify bytecodes that will initialise this data structure. When you use annotations, you can store the data directly. Moreover, annotations are typically used for metadata purposes and that intent matches this particular use case.
2) This solution works, but we want to be able to expose the debugging information in a way that libraries/tools built for Eta can make use of them in interesting ways and annotations is a great way to do that. Moreover, storing it as an attribute would require us to invest time in writing a parser for that data when rendering an extended stack trace. When done using annotations/reflection, we don't worry about parsing and can read the debugging info more or less directly via the reflection API.

**Comment by jneira** [comment]
It makes sense, thanks for the explanation. I had thought about the metadata part but to store the data directly is determinate to choose it.
